variables.tf

```
@@ -1,18 +1,98 @@
-variable "example_var" {
-  description = "Example variable passed into the module"
+variable "database_name" {
+  description = "The name of the database to create the role in"
   type        = string
 }
 
+variable "comment" {
+  description = "Database Role description"
+  type        = string
+  default     = null
+}
+
 variable "descriptor_name" {
   description = "Name of the descriptor used to form a resource name"
   type        = string
-  default     = "resource-type"
+  default     = "snowflake-role"
+}
+
+variable "parent_database_role" {
+  description = "Fully qualified Parent Database Role name (`DB_NAME.ROLE_NAME`), to create parent-child relationship"
+  type        = string
+  default     = null
+}
+
+variable "granted_database_roles" {
+  description = "Database Roles granted to this role"
+  type        = list(string)
+  default     = []
 }
 
-variable "sub_resource" {
-  description = "Some other resource that is part of stack/module"
-  type = object({
-    descriptor_name = optional(string, "sub-resource-type")
-    example_var     = string
-  })
+variable "database_grants" {
+  description = "Grants on a database level"
+  type = list(object({
+    all_privileges    = optional(bool)
+    with_grant_option = optional(bool, false)
+    privileges        = optional(list(string), null)
+    database_name     = optional(string, null)
+  }))
+  default = []
+
+  validation {
+    condition     = alltrue([for grant in var.database_grants : (grant.privileges != null) != (grant.all_privileges == true)])
+    error_message = "Variable `database_grants` fails validation - only one of `privileges` or `all_privileges` can be set."
+  }
+}
+
+variable "schema_grants" {
+  description = "Grants on a schema level"
+  type = list(object({
+    all_privileges             = optional(bool)
+    with_grant_option          = optional(bool, false)
+    privileges                 = optional(list(string), null)
+    all_schemas_in_database    = optional(bool)
+    future_schemas_in_database = optional(bool)
+    schema_name                = optional(string, null)
+  }))
+  default = []
+  validation {
+    condition     = alltrue([for grant in var.schema_grants : (grant.privileges != null) != (grant.all_privileges == true)])
+    error_message = "Variable `schema_grants` fails validation - only one of `privileges` or `all_privileges` can be set."
+  }
+  validation {
+    condition = alltrue([for grant in var.schema_grants :
```

Minor - pls fix layout:
```suggestion
      condition = alltrue([for grant in var.schema_grants :
      sum([
        grant.all_schemas_in_database != null ? 1 : 0,
        grant.future_schemas_in_database != null ? 1 : 0,
        grant.schema_name != null ? 1 : 0
      ]) == 1
    ])
```

---

The other question is - why not just use `bool` logic here?
```hcl
grant.all_schemas_in_database != null &&  grant.future_schemas_in_database != null && grant.schema_name != null
```

That would return `true` or `false`